src/graph/impl/KokkosGraph_GraphColor_impl.hpp

```diff
@@ -2379,6 +2379,300 @@ class GraphColor_VB:public GraphColor <HandleType,lno_row_view_t_,lno_nnz_view_t
+          << "\tticToc:" << (int) this->_ticToc << std::endl
+          << "\tchunkSize:" << this->_chunkSize << std::endl;
+    }
+
+    nnz_lno_t numVertices = this->nv;
+
+    size_type maxColors = 0;
+    const_lno_row_view_t myXadj = this->xadj;
+    const_lno_nnz_view_t myAdj  = this->adj;
+    nnz_lno_persistent_work_view_t score
+      = nnz_lno_persistent_work_view_t(Kokkos::ViewAllocateWithoutInitializing("score"), numVertices);
+    typedef typename Kokkos::Experimental::Max<size_type, MyExecSpace> maxScoreReducerType;
+    maxScoreReducerType maxScoreReducer(maxColors);
+    functorScoreCalculation<nnz_lno_persistent_work_view_t, size_type, MyExecSpace> scoreCalculation(score, myXadj);
+    Kokkos::parallel_reduce("Deterministic Coloring: compute initial scores", numVertices,
+                            scoreCalculation, maxScoreReducer);
+
+   if (this->_ticToc) {
+     std::cout << "maxColors: " << maxColors << std::endl;
+   }
+
+    // Create the dependency list of the graph
+    nnz_lno_persistent_work_view_t dependency
+      = nnz_lno_persistent_work_view_t("dependency", numVertices);
+    Kokkos::View<size_type, MyTempMemorySpace> frontierSize("newFrontier");
+    typename Kokkos::View<size_type, MyTempMemorySpace>::HostMirror host_frontierSize =
+      Kokkos::create_mirror_view(frontierSize);
+    Kokkos::View<size_type, MyTempMemorySpace> newFrontierSize("newFrontierSize");
+    typename Kokkos::View<size_type, MyTempMemorySpace>::HostMirror host_newFrontierSize =
+      Kokkos::create_mirror_view(newFrontierSize);
+    nnz_lno_temp_work_view_t frontier = nnz_lno_temp_work_view_t("frontier", numVertices);
+    nnz_lno_temp_work_view_t newFrontier = nnz_lno_temp_work_view_t("new frontier", numVertices);
+    Kokkos::parallel_for("Deterministic Coloring: compute dependency list", numVertices,
+                         KOKKOS_LAMBDA(const int node) {
+                           int myScore = score(node);
+                           int numNeighs = myXadj(node + 1) - myXadj(node);
+                           for(int neigh = 0; neigh < numNeighs; ++neigh) {
+                             if(myScore < score(myAdj(myXadj(node) + neigh))) {
+                               dependency(node) = dependency(node) + 1;
+                             }
+                             if(( myScore == score(myAdj(myXadj(node) + neigh)) ) &&
+                                ( node < myAdj(myXadj(node) + neigh) )) {
+                               dependency(node) = dependency(node) + 1;
+                             }
+                           }
+                           if(dependency(node) == 0) {
+                             const size_type newFrontierIdx
+                               = Kokkos::atomic_fetch_add(&newFrontierSize(), 1);
+                             newFrontier(newFrontierIdx) = node;
+                           }
+                         });
+
+    Kokkos::deep_copy(host_newFrontierSize, newFrontierSize);
```

Comment:
This is something we need to explore. It would be better to avoid this copy every time and launch another kernel later. May be just a TBD for now.

Reply:
Yes, I think that ultimately the whole while loop probably needs to be pushed on the device but I am not a big fan of writing "fake" `parallel_for` functions to push code on device although at the moment that is probably the only option?
Short of putting the while loop on the device we need to transfer some data back and forth at each iteration to know if the graph has been explored completely or if there is still a new frontier to explore.

Reply:
I wish there is a way to return a value or a device initiated copy back to host. I am fine with this for now.